server/src/main/java/io/druid/server/coordination/ZkCoordinator.java

```diff
@@ -179,74 +186,78 @@ public void addSegment(DataSegment segment, DataSegmentChangeCallback callback)
     }
   }
 
-  public void addSegments(Iterable<DataSegment> segments, final DataSegmentChangeCallback callback)
+  private void addSegments(Collection<DataSegment> segments, final DataSegmentChangeCallback callback)
```

Comment:
Given that this is a background announcer, does this need to complete atomically per callback method?

If this method itself does not need to be atomic, then it seems to me you could simply submit a listenable future to the executor service, and have success/failure callback on the futures list to handle the error reporting or DataSegmentChangeCallback. Then you don't have to worry about latches or blocking at all.